File: apps/desktop/src-tauri/src/app/commands/spellcheck.rs
//! Spell checking commands

use crate::domain::spellcheck;

/// Check if spell checking is available on this platform
#[tauri::command]
pub fn spellcheck_available() -> bool {
    spellcheck::is_available()
}

/// Check spelling of the given text
#[tauri::command]
pub fn check_spelling(text: String) -> Vec<spellcheck::SpellingError> {
    spellcheck::check_spelling(&text)
}

/// Get available languages for spell checking
#[tauri::command]
pub fn get_spell_languages() -> Vec<String> {
    spellcheck::get_available_languages()
}

/// Add a word to the user's dictionary
#[tauri::command]
pub fn learn_spelling(word: String) {
    spellcheck::learn_word(&word);
}

/// Remove a word from the user's dictionary
#[tauri::command]
pub fn unlearn_spelling(word: String) {
    spellcheck::unlearn_word(&word);
}

/// Ignore a word for this session
#[tauri::command]
pub fn ignore_spelling(word: String) {
    spellcheck::ignore_word(&word);
}

File: apps/desktop/src/services/spellcheck.ts
/**
 * Spell checking service using system APIs
 * - macOS: NSSpellChecker
 * - Windows: ISpellChecker
 * - Linux: Not supported
 */

import { invoke } from '@tauri-apps/api/core';

import { isTauri } from './index';

export interface SpellingError {
  start: number;
  length: number;
  word: string;
  suggestions: string[];
}

let availabilityCache: boolean | null = null;

/**
 * Check if spell checking is available on this platform
 */
export async function isSpellCheckAvailable(): Promise<boolean> {
  if (!isTauri()) {
    return false;
  }

  if (availabilityCache !== null) {
    return availabilityCache;
  }

  try {
    availabilityCache = await invoke<boolean>('spellcheck_available');
    return availabilityCache;
  } catch (e) {
    console.error('[SpellCheck] Failed to check availability:', e);
    availabilityCache = false;
    return false;
  }
}

/**
 * Check spelling of the given text
 */
export async function checkSpelling(text: string): Promise<SpellingError[]> {
  if (!isTauri()) {
    return [];
  }

  try {
    return await invoke<SpellingError[]>('check_spelling', { text });
  } catch (e) {
    console.error('[SpellCheck] Failed to check spelling:', e);
    return [];
  }
}

/**
 * Get available languages for spell checking
 */
export async function getSpellLanguages(): Promise<string[]> {
  if (!isTauri()) {
    return [];
  }

  try {
    return await invoke<string[]>('get_spell_languages');
  } catch (e) {
    console.error('[SpellCheck] Failed to get languages:', e);
    return [];
  }
}

/**
 * Add a word to the user's dictionary
 */
export async function learnWord(word: string): Promise<void> {
  if (!isTauri()) {
    return;
  }

  try {
    await invoke('learn_spelling', { word });
  } catch (e) {
    console.error('[SpellCheck] Failed to learn word:', e);
  }
}

/**
 * Remove a word from the user's dictionary
 */
export async function unlearnWord(word: string): Promise<void> {
  if (!isTauri()) {
    return;
  }

  try {
    await invoke('unlearn_spelling', { word });
  } catch (e) {
    console.error('[SpellCheck] Failed to unlearn word:', e);
  }
}

/**
 * Ignore a word for this session
 */
export async function ignoreWord(word: string): Promise<void> {
  if (!isTauri()) {
    return;
  }

  try {
    await invoke('ignore_spelling', { word });
  } catch (e) {
    console.error('[SpellCheck] Failed to ignore word:', e);
  }
}
